### ISSUE TITLE
Avahi compatibility libraries

### ISSUE THREAD

**Issue by marcosfrm** [systemd/systemd#21669]
`systemd-resolved` supports mDNS for some time now and already has an API to advertise services on the network (`.dnssd` files) since v236 (`man 5 systemd.dnssd`).

It would be great avoid duplication with Avahi (which is kind of unmaintained these days) having `libavahi-client.so`, `libavahi-common.so`, etc… compatible libraries.

**Comment by iam-TJ** [comment]
I'll second this. I havemulti-homed 'domain' servers with `systemd-networkd` and `systemd-resolved` doing all the heavy lifting including mDNS on IPv6-only networks.

I have currently to use a very brittle configuration where `avahi-daemon` is also configured to operate in limited ways on some interfaces (networks) alongside `systemd` in order to provide dynamic DNS-SD for several services that depend in it.

The latest to cause issues and waste time was trying to get `CUPS` **IPP Everywhere** to work on networks where `avahi` isn't allowed to operate but since `cups-daemon` depends on `avahi-client` (this is Debian 11) that really needs an avahi compatible shim or bridge.

**Comment by kepstin** [comment]
Should note that there's also a (very small) number of apps on linux which use the classic Apple mDNSResponder ["libdns_sd.so.1" library interface](https://github.com/apple-oss-distributions/mDNSResponder/blob/main/mDNSShared/dns_sd.h).

Avahi currently provides a compatibility layer for this interface, packaged as libavahi-compat-dnssd1 on debian. It would probably make sense for an Avahi compatibility library project to also provide the mDNSResponder compatibility library.

There's only a couple debian packages that use it, tho, so it's not a very high priority:
```
  pike8.0-dnssd
  mumble-server
  mumble
  barrier
  aseba
```

**Comment by LorbusChris** [comment]
With systemd v259, sd-resolved can replace Avahi for mDNS discovery and announcement.

I've built a PoC compat lib/drop in header replacement: https://github.com/LorbusChris/resolve-avahi-compat. Only the browsing/discovery part is currently tested with that. It also requires https://github.com/systemd/systemd/pull/40133

If you only need to browse on a specific network interface, or you iterate through them in your codebase, the new varlink BrowseServices API in v259 can just be used as-is. Here's an example implementation: https://gitlab.gnome.org/GNOME/gnome-network-displays/-/blob/master/src/nd-sd-wfd-mice-provider.c

**Comment by niklaswimmer** [comment]
Would it also be necessary to provide a replacement for the avahi dbus endpoint `org.freedesktop.Avahi`? I do not know if there are any applications out there that skipped `libavahi-client` and just called the dbus endpoint directly, it seems not unlikely though?

How would such a migration look like?

**Comment by travisghansen** [comment]
Yeah good point. I know I have ran into those before but don’t recall the specific apps atm.
